src/python/pants/util/strutil.py

```diff
@@ -335,3 +339,31 @@ def wrapper(func: Callable[P, R]) -> Callable[P, R]:
         return func
 
     return wrapper
+
+
+class _JsonEncoder(json.JSONEncoder):
+    """Allow us to serialize everything, with a fallback on `str()` in case of any esoteric
+    types."""
+
+    def default(self, o):
+        """Return a serializable object for o."""
+        if dataclasses.is_dataclass(o):
+            return dataclasses.asdict(o)
+        if isinstance(o, abc.Mapping):
+            return dict(o)
+        if isinstance(o, abc.Sequence):
+            return list(o)
+        try:
+            return super().default(o)
+        except TypeError:
+            return str(o)
```

Comment:
It would probably be better not to fall back to `str`, as that could mask valid errors... but not sure if @kaos added that for a particular case he was seeing.

Reply:
what valid error would that be?

My intention here is that what ever value I pass in, I want it to ~never fail. You should be able to pass in a `Version` instance for example, but that doesn't come with a json encoder so will throw the `TypeError`.

Reply:
Maybe we should do the safe thing first and just error?

Reply:
> what valid error would that be?

This is a genuine question. And if we can't come up with an answer to it, I see no reason to let it propagate up the stack.

Reply:
OK, that's fair. My wish for it to never throw stands though, and for the types I've seen the most you get a `<ClassName possible=some attributes>` styled string back for `str`.. maybe `repr` would work better as that is intended to be the source code equiv of the instance that would reproduce if evaluated, when possible.

I would be ok with an option to `stable_hash` which selects if a fallback to `repr` (or `str`) is preferred over raising an error.